s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/S3ObjectSummaryIterator.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.s3.source.utils;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * Implements a ObjectSummaryIterator on an S3 bucket. Implementation reads summaries in blocks and iterates over each
+ * block. When block is empty a new block is retrieved and processed until no more data is available.
+ */
+public class S3ObjectSummaryIterator implements Iterator<S3ObjectSummary> {
+    /** The client we are using */
+    private final AmazonS3 s3Client;
+    /** The object listing from the last call to the client */
+    private ListObjectsV2Result objectListing;
+    /** The inner iterator on the object summaries. When it is empty a new one is read from object listing. */
+    private Iterator<S3ObjectSummary> innerIterator;
+
+    /** the ObjectRequest initially to start the iteration from later to retrieve more records */
+    private final ListObjectsV2Request request;
+
+    /** The last key seen by this process. This allows us to restart when a new file is dropped in the direcotry */
```

Comment:
I would say not necessarily when a new file is dropped, more like just let us query S3 from after the last known processed file.

Reply:
maybe a nit, I just think wording is misleading.